### **Assignment 1: Finding the Best Delivery Promise for Rosa's Pizza**


**This notebook analyzes Rosa's delivery data to measure how often orders are late (Part I), find the most profitable delivery promise for each zone and time block (Part II), and support a web app for Rosa (Part III).**

In [10]:
!pip install -q git+https://github.com/zhouy185/rosa-starter.git

  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... done


**Setup:** Install the starter package, import the data and functions, and check that the simulator works. All calls use `seed=1` so results are the same every run.


In [2]:
import numpy as np
from starter import ZONES, TIME_BLOCKS, COSTS, PROMISE, delivery_times

In [11]:
times = delivery_times("Far West", "Fri/Sat eve", 45, seed=1)
print(len(times), "orders; average delivery time", round(times.mean(), 1),
"minutes")

210 orders; average delivery time 42.9 minutes


### **How Bad Is It?**

### **Part I: (a)**
### **Percentage of Late Orders**

In [12]:

import numpy as np
from starter import ZONES, TIME_BLOCKS, delivery_times

def get_late_percentage(zone, time_block, promise, seed=1):

    # CASE 1: Both zone and time block are specific names
    if zone != 'all' and time_block != 'all':
        times = delivery_times(zone, time_block, promise, seed=1)
        return (times > promise).mean() * 100

    # CASE 2: Zone is 'all', but time block is specific
    if zone == 'all' and time_block != 'all':
        basket = []
        for z in ZONES:
            times = delivery_times(z, time_block, promise, seed=1)
            basket.append(times)

        data = np.concatenate(basket)
        return (data > promise).mean() * 100

    # CASE 3: Specific zone, but time block is 'all'
    if zone != 'all' and time_block == 'all':
        basket = []
        for tb in TIME_BLOCKS:
            times = delivery_times(zone, tb, promise, seed=1)
            basket.append(times)

        data = np.concatenate(basket)
        return (data > promise).mean() * 100

    # CASE 4: Both zone and time block are 'all'
    if zone == 'all' and time_block == 'all':
        basket = []
        for z in ZONES:
            for tb in TIME_BLOCKS:
                times = delivery_times(z, tb, promise, seed=1)
                basket.append(times)

        data = np.concatenate(basket)
        return (data > promise).mean() * 100


# Test the function

# Test Case 1
print("Case 1:", get_late_percentage("Far West", "Fri/Sat eve", 45))

# Test Case 2
print("Case 2:", get_late_percentage('all', "Lunch", 45))

# Test Case 3
print("Case 3:", get_late_percentage("North", 'all', 45))

# Test Case 4
print("Case 4:", get_late_percentage('all', 'all', 45))

Case 1: 36.666666666666664
Case 2: 3.3333333333333335
Case 3: 6.800000000000001
Case 4: 8.291873963515755


### **Part I: (b)**
### **Ranking Zones and Time Blocks by Late Rate**

In [13]:
zone_ranking = []

# Loop through every zone and every time block
for z in ZONES:
    for tb in TIME_BLOCKS:
        # Call the function I wrote with a 45-minute promise
        late_rate = get_late_percentage(z, tb, 45)

        # Save the zone, time block, and its late rate into my list
        zone_ranking.append((z, tb, late_rate))

# Sort the list from highest late rate to lowest (descending order)
zone_ranking_sorted = sorted(zone_ranking, key=lambda x: x[2], reverse=True)

# Print out the results
print("Ranked Late Arrival Rates (descending order):")
for zone, time_block, rate in zone_ranking_sorted:
    print("Zone: " + zone + " - Time Block: " + time_block + " - Late Rate: " + str(rate) + "%")

Ranked Late Arrival Rates (descending order):
Zone: Far West - Time Block: Fri/Sat eve - Late Rate: 36.666666666666664%
Zone: North - Time Block: Fri/Sat eve - Late Rate: 16.666666666666664%
Zone: Far West - Time Block: Weekday eve - Late Rate: 11.176470588235295%
Zone: Central - Time Block: Fri/Sat eve - Late Rate: 6.842105263157896%
Zone: North - Time Block: Weekday eve - Late Rate: 6.0%
Zone: Far West - Time Block: Other - Late Rate: 4.6875%
Zone: Central - Time Block: Weekday eve - Late Rate: 4.242424242424243%
Zone: North - Time Block: Lunch - Late Rate: 3.684210526315789%
Zone: Central - Time Block: Lunch - Late Rate: 3.4615384615384617%
Zone: Central - Time Block: Other - Late Rate: 2.631578947368421%
Zone: North - Time Block: Other - Late Rate: 2.5%
Zone: Far West - Time Block: Lunch - Late Rate: 2.5%


### **Part I: (c)**
### **Average Delivery Time**

In [14]:
def get_avg_delivery_time(zone, time_block, promise, seed=1):

    # CASE 1: Both zone and time block are specific names
    if zone != 'all' and time_block != 'all':
        times = delivery_times(zone, time_block, promise, seed=1)
        return times.mean()

    # CASE 2: Zone is 'all', but time block is specific
    if zone == 'all' and time_block != 'all':
        basket = []
        for z in ZONES:
            times = delivery_times(z, time_block, promise, seed=1)
            basket.append(times)

        data = np.concatenate(basket)
        return data.mean()

    # CASE 3: Specific zone, but time block is 'all'
    if zone != 'all' and time_block == 'all':
        basket = []
        for tb in TIME_BLOCKS:
            times = delivery_times(zone, tb, promise, seed=1)
            basket.append(times)

        data = np.concatenate(basket)
        return data.mean()

    # CASE 4: Both zone and time block are 'all'
    if zone == 'all' and time_block == 'all':
        basket = []
        for z in ZONES:
            for tb in TIME_BLOCKS:
                times = delivery_times(z, tb, promise, seed=1)
                basket.append(times)

        data = np.concatenate(basket)
        return data.mean()

# Testing the function

# Test Case 1
print("Case 1:", get_avg_delivery_time("Far West", "Fri/Sat eve", 45))

# Test Case 2
print("Case 2:", get_avg_delivery_time('all', "Lunch", 45))

# Test Case 3
print("Case 3:", get_avg_delivery_time("North", 'all', 45))

# Test Case 4
print("Case 4:", get_avg_delivery_time('all', 'all', 45))

Case 1: 42.9195238095238
Case 2: 28.219298245614034
Case 3: 30.9752
Case 4: 31.193739635157545


### **Part I: (d)**
### **Ranking Zones and Time Blocks by Average Delivery Time**

In [15]:
zone_ranking_avg_delivery = []

# Loop through every zone and every time block
for z in ZONES:
    for tb in TIME_BLOCKS:
        avg_time = get_avg_delivery_time(z, tb, 45)

        # Saving the zone, time block, and its average delivery time into our list
        zone_ranking_avg_delivery.append((z, tb, avg_time))

# Sort the list from highest average delivery time to lowest (descending order)
zone_ranking_avg_delivery_sorted = sorted(zone_ranking_avg_delivery, key=lambda x: x[2], reverse=True)

# Print the result
print("Ranked Average Delivery Times (descending order):")
for zone, time_block, avg_time in zone_ranking_avg_delivery_sorted:
    print("Zone: " + zone + " - Time Block: " + time_block + " - Avg Time: " + str(avg_time) + " minutes")

Ranked Average Delivery Times (descending order):
Zone: Far West - Time Block: Fri/Sat eve - Avg Time: 42.9195238095238 minutes
Zone: North - Time Block: Fri/Sat eve - Avg Time: 36.76533333333334 minutes
Zone: Far West - Time Block: Weekday eve - Avg Time: 35.044117647058826 minutes
Zone: Central - Time Block: Fri/Sat eve - Avg Time: 32.13263157894737 minutes
Zone: North - Time Block: Weekday eve - Avg Time: 31.3756 minutes
Zone: Far West - Time Block: Lunch - Avg Time: 30.58083333333333 minutes
Zone: Far West - Time Block: Other - Avg Time: 30.04010416666667 minutes
Zone: Central - Time Block: Weekday eve - Avg Time: 29.030909090909088 minutes
Zone: North - Time Block: Lunch - Avg Time: 28.672631578947364 minutes
Zone: North - Time Block: Other - Avg Time: 27.655624999999997 minutes
Zone: Central - Time Block: Lunch - Avg Time: 26.798076923076923 minutes
Zone: Central - Time Block: Other - Avg Time: 26.03684210526316 minutes


All the averages (26mins - 43 mins) lie under the 45mins promise which is why it is difficult to find why the orders might be late.

### **Part I: (e)**: **Is late rate or Avg Delivery Time measure better for Rosa's business decision making?**

The late rate (b) is more relevant to Rosa's decisions. Based on our analysis, the longest average delivery time is 42.9 minutes, which is technically still under the 45-minute promise. However, that 42.9-minute average also means that 36.7% of her customers are getting their orders late, which is quite a lot. From a customer's perspective, the late rate measures how often Rosa breaks her brand promise, making it a core metric for business decisions.


### **PART II: Exploring The Best Promise**

### **Part II: (a)**
### **Total Cost of Late Order**

In [16]:
def calculate_late_cost(costs):
    refund_cost = costs["refund"]

    # Churn cost: future lost orders multiplied by profit margin per order
    churn_cost = costs["churn_orders"] * costs["margin"]

    total_cost = refund_cost + churn_cost
    return total_cost

# Test the function using the COSTS dictionary and printing the output
cost_per_late = calculate_late_cost(COSTS)
print("Total cost per late order: $" + str(cost_per_late))

Total cost per late order: $26.2


Each late order costs 10 dollars in refunds plus 1.8 lost future orders * 9 dollars margin, for a total of $26.20.

### **Part II: (b)**: **Choosing the right range of Promises**

If we were to consider a range from 5 to 60 minutes in increments of 5 minutes, 10 to 25 minutes would be too little time for delivery. Even though a shorter delivery time seems very appealing to customers, Rosa's business would suffer deeply: order volumes exceeding the pizzeria's capacity would result in massive complaints, refund costs, and customer churn.

Choosing a range from 30 to 90 minutes in 5-minute increments is the best option because it allows Rosa to find the exact balance, maintaining a high volume of customer orders to generate revenue while keeping promises tight enough to control late-arrival refund requests. For a range to be good, the best promise must fall inside it and not at the extremes of it.

Conversely, anything over 60 minutes would ensure on-time delivery and little-to-no refund or churn costs, but long delivery times would make the pizzeria unappealing to customers, dropping order volume. We still include 65 to 90 minutes to confirm that profit keeps falling past the peak.

The results confirm the range is right: net profit rises, peaks at 55 minutes ($1,212.40), and then falls.

### **Function to find the Best Promise**

In [22]:
def find_best_promise(zone, time_block, promise_list, costs, seed=1):
    # Calculate total cost per late order
    cost_per_late = calculate_late_cost(costs)

    best_promise = None
    max_net_profit = -999999.0

    for p in promise_list:
        # Simulate delivery times for this specific promise
        times = delivery_times(zone, time_block, promise=p, seed=1)
        n_orders = len(times)

        if n_orders == 0:
            net_profit = 0.0
        else:
            # Count how many orders arrived after the promised time (late)
            n_late = sum(1 for t in times if t > p)

            # Gross profit collected from all orders
            gross_profit = n_orders * costs["margin"]

            # Total late costs incurred
            total_late_cost = n_late * cost_per_late

            # Net profit = Gross profit minus late costs
            net_profit = gross_profit - total_late_cost

        # Track the promise that yields the highest net profit
        if net_profit > max_net_profit:
            max_net_profit = net_profit
            best_promise = p

    return best_promise, max_net_profit


# Testing the function

# Define the range of promises to test in 5-minute steps
promises_to_try = [30, 35, 40, 45, 50, 55, 60, 65, 70, 75, 80, 85, 90]

# Test the function on "Far West" and "Fri/Sat eve"
zone_test = "Far West"
time_block_test = "Fri/Sat eve"

best_p, profit = find_best_promise(zone_test, time_block_test, promises_to_try, COSTS)

# Printing the output
print("Recommended Promise for " + zone_test + " (" + time_block_test + "): " + str(best_p) + " minutes")
print("Expected Net Profit: $" + str(round(profit, 2)))

cost_per_late = calculate_late_cost(COSTS)

# Net profit at each promise for Far West, Fri/Sat eve
for p in promises_to_try:
    times = delivery_times("Far West", "Fri/Sat eve", p, seed=1)
    n_late = (times > p).sum()
    late_rate = n_late / len(times) * 100
    net_profit = len(times) * COSTS["margin"] - n_late * cost_per_late
    print("Promise: " + str(p) + " min - Orders: " + str(len(times)) + " - Late Rate: " + str(round(late_rate, 1)) + "% - Net Profit: $" + str(round(net_profit, 2)))

    # Best promise for every zone and time block
for z in ZONES:
    for tb in TIME_BLOCKS:
        best_p, profit = find_best_promise(z, tb, promises_to_try, COSTS)
        print("Zone: " + z + " - Time Block: " + tb + " - Best Promise: " + str(best_p) + " min - Net Profit: $" + str(round(profit, 2)))


Recommended Promise for Far West (Fri/Sat eve): 55 minutes
Expected Net Profit: $1212.4
Promise: 30 min - Orders: 237 - Late Rate: 92.0% - Net Profit: $-3578.6
Promise: 35 min - Orders: 232 - Late Rate: 77.2% - Net Profit: $-2601.8
Promise: 40 min - Orders: 224 - Late Rate: 59.4% - Net Profit: $-1468.6
Promise: 45 min - Orders: 210 - Late Rate: 36.7% - Net Profit: $-127.4
Promise: 50 min - Orders: 188 - Late Rate: 17.6% - Net Profit: $827.4
Promise: 55 min - Orders: 158 - Late Rate: 5.1% - Net Profit: $1212.4
Promise: 60 min - Orders: 121 - Late Rate: 1.7% - Net Profit: $1036.6
Promise: 65 min - Orders: 84 - Late Rate: 0.0% - Net Profit: $756.0
Promise: 70 min - Orders: 54 - Late Rate: 0.0% - Net Profit: $486.0
Promise: 75 min - Orders: 32 - Late Rate: 0.0% - Net Profit: $288.0
Promise: 80 min - Orders: 18 - Late Rate: 0.0% - Net Profit: $162.0
Promise: 85 min - Orders: 10 - Late Rate: 0.0% - Net Profit: $90.0
Promise: 90 min - Orders: 6 - Late Rate: 0.0% - Net Profit: $54.0
Zone: Cent

### **Best Promise for every Zone and time Block**

Most zones and time blocks are already doing best at 45 minutes, so Rosa doesn't need to change everything. The promise only needs to go up in three places: Far West on Fri/Sat evenings (55 min), North on Fri/Sat evenings (50 min), and Far West on weekday evenings (50 min). Central on Fri/Sat evenings is also best at 50, but it only adds about $30, so it's not worth changing

### **Recommendation for Rosa**

Far West on Fri/Sat evenings is the biggest problem. At today's 45-minute promise, Rosa is actually losing money there (–$127.40) because more than a third of orders arrive late. Moving to 55 minutes cuts late orders to about 5% and brings profit up to about 1,212 dollars, a gain of roughly 1,340 dollars over four weeks.

### **PART III: Creating a Web App**

The web page that will help Rosa choose the most profitable delivery promise for any zone and time block. Rosa can manipulate the range of promised times and adjust the costs to see the recommendations.

Github: https://github.com/mannalsheikh/rosa-pizza-promise-app

Streamlit app: https://rosa-pizza-promise-app-cghr4g84kumcvh67fh6ys9.streamlit.app/

### **AI Appendix:**

I used Claude mainly to understand, write and fix code. My part was working out what Rosa's problem needed, explaining it to Claude, and checking that the results made sense.

**Part I:** I wrote the functions using examples from class, but they gave an error when the zone or time block was "all." Claude fixed how my code combined the delivery times. I then tested all four cases.

**Part II:** I explained the calculation (refund plus lost future orders per late order, and choosing the promise with the highest profit), and Claude wrote the two functions. I checked the cost by hand: $10 + 1.8 × $9 = $26.20.

**Extra Part II cells:** While I was finishing Part II, Claude suggested two extra cells. One shows the profit at every promise, and the other finds the best promise for every zone and time block. Claude wrote the code and helped me word the explanations after I interpreted the results.

**Part III:** I asked Claude to create the notebook-to-streamlit skill, a set of instructions for turning my notebook into an app without changing my formulas. The skill lists six steps (read the notebook, copy the functions, build the app, write the requirements, test it, and deploy it) and rules such as keeping seed=1 so the app matches my notebook. Following the skill, Claude copied my two functions into rosa_logic.py (because Streamlit can't read a notebook directly) and built app.py, requirements.txt and CLAUDE.md.

**Prompts I used:**
Part I: "I'm trying to write the function for Part I where either the zone, the time block, or both can be set to 'all'. My current draft is throwing an error. Can you help fix my code so it calculates the correct late percentages and averages?"

**Part II:** "Here is the workflow for Rosa's analysis logic: 1. Write calculate_late_cost(costs) to compute direct refund costs plus churn costs (future lost orders multiplied by the profit margin per order). 2. Write find_best_promise(zone, time_block, promise_list, costs, seed=1) to loop through promised delivery times in 5-minute steps, simulate delivery times using delivery_times, calculate gross profit minus total late costs, and return the optimal promise and maximum net profit. Please generate clean, commented Python code following this exact logic."

**Part III** (reconstructed from memory, since I accidentally deleted this conversation): "I've opened my notebook. Can you create a skill called notebook-to-streamlit with instructions for turning my notebook into a Streamlit app without changing my formulas? Then use it to build the app for Rosa, and also make a CLAUDE.md file and anything else the app needs to run." Claude then gave me a command to run in the terminal, which opened the app in my browser so I could test it before putting it on Streamlit Community Cloud. The skill file is in my repository at .github/skills/notebook-to-streamlit/SKILL.md.

